### ISSUE TITLE
Any plan for utterance batch decoding for Transformer asr model?

### ISSUE THREAD

**Issue by keshawnhsieh** [espnet/espnet#2679]
I noticed that RNN based model has the **utterance batch** decoding method implemented in [recognize_batch](https://github.com/espnet/espnet/blob/5d97a488c52cd303993064e147fb3a2be20337a5/espnet/nets/pytorch_backend/e2e_asr.py#L394), however there is no such implementation for Transformer based asr model.

I understand that more advanced interface for beam search has been moved to [recog_v2](https://github.com/espnet/espnet/blob/5d97a488c52cd303993064e147fb3a2be20337a5/espnet/asr/pytorch_backend/recog.py#L22), but in this new implementation  I only can find the **beam batch** interface [batch_score](https://github.com/espnet/espnet/blob/5d97a488c52cd303993064e147fb3a2be20337a5/espnet/nets/pytorch_backend/transformer/decoder.py#L335) rather than utterance batch decode method. So any plan for the utterance batch decoding for Transformer based asr model?

**Comment by sw005320** [comment]
This is a very good discussion.
We decided to work on the batch scoring only and did not implement an utterance batch since it makes our implementation simple for the API v2 and also the main ASR decoding scenario uses a single utterance.
However, we also think the utterance batch is important for the off-line ASR scenarios and decoding during training.
We can make this part a higher priority upon a request.
(Added @ShigekiKarita for this discussion.)

**Comment by ShigekiKarita** [comment]
I wonder if we can use multi processing API that requires no changes in recog_v2 https://pytorch.org/docs/stable/notes/multiprocessing.html
```python
import torch.multiprocessing as mp
from model import MyModel
from utterance import utt_loader

def recog(model):
    result = []
    for i, utt in utt_loader('path to utt list'):
        if i % num_processes == rank:
            result.append(model.recognize(utt))
    torch.save(result, f'result.{rank}.pkl')

if __name__ == '__main__':
    num_processes = 4
    model = MyModel('path to model')
    processes = []
    for rank in range(num_processes):
        p = mp.Process(target=recog, args=(model,))
        p.start()
        processes.append(p)
    for p in processes:
        p.join()
```

**Comment by keshawnhsieh** [comment]
> I wonder if we can use multi processing API that requires no changes in recog_v2 https://pytorch.org/docs/stable/notes/multiprocessing.html
> 
> ```python
> import torch.multiprocessing as mp
> from model import MyModel
> from utterance import utt_loader
> 
> def recog(model):
>     result = []
>     for utt in utt_loader('path to utt list'):
>         result.append(model.recognize(utt))
>     torch.save(result, f'result.{rank}.pkl')
> 
> if __name__ == '__main__':
>     num_processes = 4
>     model = MyModel('path to model')
>     processes = []
>     for rank in range(num_processes):
>         p = mp.Process(target=recog, args=(model,))
>         p.start()
>         processes.append(p)
>     for p in processes:
>         p.join()
> ```

**Comment by keshawnhsieh** [comment]
Maybe it's a not a good ideas to archive such purpose by this way. During Decoding, I observed a strange phenomenon that GPU-Util reached at a high level like 90%+  while the true power usage was much lower. Shown in below.
<img width="648" alt="Snipaste_2020-11-17_11-15-27" src="https://user-images.githubusercontent.com/29997158/99342634-4cfa3100-28c7-11eb-81ff-5db18f5ca885.png">

I don't know much about GPU working mechanism, but according to my experiment experience, we will not get much performance gain in such situation. However, merging utt batch into beam batch may help.

**Comment by sw005320** [comment]
Year, @ShigekiKarita's solution does not solve the GPU efficiency issue, but it is useful to extend it without changing our API.

What is your purpose?
Our current implementation corresponds to the S=1 result at Table 6 in https://www.isca-speech.org/archive/Interspeech_2019/pdfs/2860.pdf
Yes, if we use utterance batch, we could obtain around two times faster decoding but the normal test set is much smaller than the training data and we also don't need multiple epochs. 
So, I don't have so many issues with the current implementation.

If the purpose is to perform decoding during training (e.g., MBR), it would be a problem.
Or if your application is to automatically transcribe a huge amount of speech data, it would have great benefits of using utterance batch.
We also have such use cases and I can increase the priority of the development of utterance batch processing depending on your request.

**Comment by hchung12** [comment]
Utterance batch decoding is very useful function for semi-supervised training using large unlabeled speech corpus, especially for self-training or pseudo labeling methods. So, it would be nice to support utterance batch decoding.

**Comment by sw005320** [comment]
Yes, this is another super important application. Thanks for your comment, @hchung12.
We'll make the priority of this development item higher.

**Comment by keshawnhsieh** [comment]
> Year, @ShigekiKarita's solution does not solve the GPU efficiency issue, but it is useful to extend it without changing our API.
> 
> What is your purpose?
> Our current implementation corresponds to the S=1 result at Table 6 in https://www.isca-speech.org/archive/Interspeech_2019/pdfs/2860.pdf
> Yes, if we use utterance batch, we could obtain around two times faster decoding but the normal test set is much smaller than the training data and we also don't need multiple epochs.
> So, I don't have so many issues with the current implementation.
> 
> If the purpose is to perform decoding during training (e.g., MBR), it would be a problem.
> Or if your application is to automatically transcribe a huge amount of speech data, it would have great benefits of using utterance batch.
> We also have such use cases and I can increase the priority of the development of utterance batch processing depending on your request.

Thanks. To be honest, my application might be more engineering as there will be multiple audio utterances need to be processed in the same time and it's a time sensitive situation. Using torch.multiprocessing may not be a good idea because GPU utilization has reached its bottleneck while doing utterance batching can further improve the decoding speed (I guess).

I think @hchung12 really provided a great situation that utt batch decoding can help a lot and such situation is more common in research purpose.

Thank you for all your helpful discussions and I will close this issue at this moment.
